doc: add some bits about demand paging

dcpleung opened the pull request:
This adds some information about demand paging in Zephyr.

dcpleung committed:
doc: add some bits about demand paging

This adds some information about demand paging in Zephyr.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: mmu: add doxygen group for demand paging APIs

This adds the doxygen group for demand paging APIs in
sys/mem_manage.h header file.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented:
avoid creating new guides, put alongside user mode documentation under a  new section for Memory Management instead, basically move `doc/reference/usermode/` and this new section under `doc/reference/memory_management`. We are trying to clean up the guides section and consolidate documentation in one place.

dcpleung committed:
kernel: mmu: rename z_eviction* to k_mem_paging_eviction*

These functions and data structures are those that need
to be implemented by eviction algorithm and application
outside kernel. Promote them from z_* so these can be
included in documentation.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: mmu: z_backing_store* to k_mem_paging_backing_store*

These functions are those that need be implemented by backing
store outside kernel. Promote them from z_* so these can be
included in documentation.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Usermode is not exactly memory management so I instead opt to create a new page under reference.

enjiamai reviewed (approved):
Seems like a rebase is needed, one tiny conflict in the doc.
